model/src/main/kotlin/spdx/SpdxExpressionDefaultVisitor.kt

```diff
@@ -0,0 +1,88 @@
+        return when (ctx.childCount) {
+            2 -> visit(ctx.getChild(0))
+            else -> throw Error("SpdxExpression has invalid amount of children: '${ctx.childCount}'")
+        }
+    }
+
+    override fun visitCompoundExpression(ctx: SpdxExpressionParser.CompoundExpressionContext): SpdxExpression {
+        return when (ctx.childCount) {
+            1 -> visit(ctx.getChild(0))
+            3 -> {
+                if (ctx.getChild(0).text == "(" && ctx.getChild(2).text == ")") {
+                    visit(ctx.getChild(1))
+                } else {
+                    val left = visit(ctx.getChild(0))
+
+                    val operatorName = ctx.getChild(1).text
+                    val operator = try {
+                        SpdxOperator.valueOf(operatorName.toUpperCase())
+                    } catch (e: IllegalArgumentException) {
+                        throw Error("Illegal operator '$operatorName' in expression '${ctx.text}'.")
+                    }
+
+                    val right = visit(ctx.getChild(2))
+
+                    if (operator == SpdxOperator.WITH && right !is SpdxLicenseExceptionExpression) {
+                        throw Error("Argument '$right' for WITH is not an SPDX license exception id in '${ctx.text}'.")
+                    }
+
+                    SpdxCompoundExpression(left, operator, right)
+                }
+            }
+            else -> throw Error("SpdxCompoundExpression has invalid amount of children: '${ctx.childCount}'")
+        }
+    }
+
+    override fun visitSimpleExpression(ctx: SpdxExpressionParser.SimpleExpressionContext): SpdxExpression {
+        return when (ctx.childCount) {
+            1 -> visit(ctx.getChild(0))
+            else -> throw Error("SpdxSimpleExpression has invalid amount of children: '${ctx.childCount}'")
+        }
+    }
+
+    override fun visitLicenseExceptionExpression(ctx: SpdxExpressionParser.LicenseExceptionExpressionContext)
+            : SpdxExpression {
+        return when (ctx.childCount) {
+            1 -> SpdxLicenseExceptionExpression(ctx.text)
+            else -> throw Error("SpdxLicenseExceptionExpression has invalid amount of children: '${ctx.childCount}'")
+        }
+    }
+
+    override fun visitLicenseIdExpression(ctx: SpdxExpressionParser.LicenseIdExpressionContext): SpdxExpression {
+        return when (ctx.childCount) {
+            1 -> SpdxLicenseIdExpression(ctx.text)
+            2 -> SpdxLicenseIdExpression(ctx.text.dropLast(1), anyLaterVersion = true)
+            else -> throw Error("SpdxLicenseIdExpression has invalid amount of children: '${ctx.childCount}'")
+        }
+    }
+
+    override fun visitLicenseRefExpression(ctx: SpdxExpressionParser.LicenseRefExpressionContext): SpdxExpression {
+        return when (ctx.childCount) {
+            1 -> SpdxLicenseRefExpression(ctx.text)
+            else -> throw Error("SpdxLicenseRefExpression has invalid amount of children: '${ctx.childCount}'")
```

Comment:
Is this quite generic type thrown out of your new component or is it always caught so that client code won't ever get it.
If it goes outside of the component I'd vote for introducing a new throwable type to allow for more fine grained handling.

Reply:
Done. Added `SpdxException`.